compiler/rustc_parse/src/parser/item.rs

```diff
@@ -550,52 +552,103 @@ impl<'a> Parser<'a> {
-                .span_suggestion(
+        let mut ty_first =
+            if self.token.is_keyword(kw::For) && self.look_ahead(1, |t| t != &token::Lt) {
+                let span = self.prev_token.span.between(self.token.span);
+                self.struct_span_err(span, "missing trait in a trait impl")
+                    .span_suggestion(
+                        span,
+                        "add a trait here",
+                        " Trait ",
+                        Applicability::HasPlaceholders,
+                    )
+                    .span_suggestion(
+                        span.to(self.token.span),
+                        "for an inherent impl, drop this `for`",
+                        "",
+                        Applicability::MaybeIncorrect,
+                    )
+                    .emit();
+                P(Ty {
+                    kind: TyKind::Path(None, err_path(span)),
                     span,
-                    "add a trait here",
-                    " Trait ",
-                    Applicability::HasPlaceholders,
-                )
-                .span_suggestion(
-                    span.to(self.token.span),
-                    "for an inherent impl, drop this `for`",
-                    "",
-                    Applicability::MaybeIncorrect,
-                )
-                .emit();
-            P(Ty {
-                kind: TyKind::Path(None, err_path(span)),
-                span,
-                id: DUMMY_NODE_ID,
-                tokens: None,
-            })
-        } else {
-            self.parse_ty_with_generics_recovery(&generics)?
-        };
+                    id: DUMMY_NODE_ID,
+                    tokens: None,
+                })
+            } else {
+                self.parse_ty_with_generics_recovery(&generics)?
+            };
 
         // If `for` is missing we try to recover.
         let has_for = self.eat_keyword(kw::For);
         let missing_for_span = self.prev_token.span.between(self.token.span);
 
-        let ty_second = if self.token == token::DotDot {
+        let mut ty_second = if self.token == token::DotDot {
             // We need to report this error after `cfg` expansion for compatibility reasons
             self.bump(); // `..`, do not add it to expected tokens
             Some(self.mk_ty(self.prev_token.span, TyKind::Err))
         } else if has_for || self.token.can_begin_type() {
-            Some(self.parse_ty()?)
+            snapshot_before_last_ty = self.create_snapshot_for_diagnostic();
+            Some(self.parse_ty_no_anon_recovery()?)
         } else {
             None
         };
 
         generics.where_clause = self.parse_where_clause()?;
 
-        let impl_items = self.parse_item_list(attrs, |p| p.parse_impl_item(ForceCollect::No))?;
+        let (mut impl_items, err) =
+            self.parse_item_list(attrs, |p| p.parse_impl_item(ForceCollect::No))?;
+
+        if let Some(mut err) = err {
```

Comment:
Could you add a comment explaining what happens here?

Reply:
Uhh this is with the other comment that I'm not too sure if this is the right way to do things, but basically I'm trying to parse an impl block, if that fails I try and parse an anonymous struct or union type and then parse an impl block. (Really I would need to do it three times to recover form `impl union {}`, but again I'm not even sure this is the right path.

Reply:
It's less that error message is bad and it's more that I can't do recovery (at least how I'm doing it) because calling the method during recovery will emit the error without allowing me to suppress it.